File: Python/Insertion Sort List.py
# Sort a linked list using insertion sort

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None or head.next == None:
            return head
        
        node = head;
        
        while node.next != None:
            cur = node.next
            
            if cur.val >= node.val:
                node = node.next
                continue
            
            node.next = cur.next
            
            if cur.val <= head.val:
                cur.next = head
                head = cur
                continue
            
            pos = self.findInsertPos(head, cur) 
            
            cur.next = pos.next
            pos.next = cur3
            
        return head
        
        
    def findInsertPos(self, head, node):
        pos = head
        while head != None and node.val > head.val:
            pos = head
            head = head.next
        return pos

File: Java/Insertion_Sort_List.java
class Insertion_Sort_List{
 	public ListNode insertionSortList(ListNode head) {
        if(head == null || head.next == null)
        	return head;
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        dummy.next = head;
        ListNode curr = head.next;
        ListNode prev = head;
        while(curr != null){
        	if(curr.val >= prev.val){
        		prev = curr;
        		curr = curr.next;
        		continue;
        	}
        	ListNode p = dummy;
        	ListNode temp = curr.next;
        	while(p != prev){
        		if(p.val <=curr.val && p.next.val >= curr.val){
        			curr.next = p.next;
        			p.next = curr;
        			curr = temp;
        			prev.next = temp;
        			break;
        		}
        		p = p.next;
        	}
        }
        return dummy.next;
    }
}